src/dspbase.jl

```diff
@@ -718,9 +718,14 @@ function conv(u::AbstractArray{<:BLAS.BlasFloat, N},
     conv(u, float(v))
 end
 
-function conv(A::AbstractArray, B::AbstractArray)
-    maxnd = max(ndims(A), ndims(B))
-    return conv(cat(A, dims=maxnd), cat(B, dims=maxnd))
+function conv(A::AbstractArray{<:Number, M},
+              B::AbstractArray{<:Number, N}) where {M, N}
+    if (M < N)
+        conv(cat(A, dims=N)::AbstractArray{eltype(A), N}, B)
+    else
+        @assert M > N
```

Comment:
Is it your intent here to stop infinite recursion when there is no other applicable `conv` methods (as would happen with arrays of `Gray` values)?

Reply:
My intent is to ensure all methods of `conv` reject arrays of non-`Number` eltype.  Since `Gray` is not `Number`, my version will raise `MethodError` before executing any method of `conv`.

What martinholters points out as a problem I believe is what would happen if `Gray` were actually made a subtype of `Number`, but `float(::Gray{<:Number})` remained unchanged (still returning a `Gray{<:AbstractFloat}`).

Reply:
```
@assert M > N
```
should always succeed, since if `M == N` it would be some above-defined method that got called.  I put it here essentially as a comment, and also so that debugging would be easier should other methods would get messed up in the future.

Reply:
Hm, is there any chance some code defined `conv(u::AbstractArray{<:Foo, N}, v::AbstractArray{<:Foo, N}) where {N}` for some non-`Number` `Foo` and relied on the method herein to get arguments to the same dimensionality first?

In general, this method really doesn't need its argument eltypes to be restricted to `Number`. Maybe a better approach would be to define `conv(u::AbstractArray{<:Any, N}, v::AbstractArray{<:Any, N}) where {N}` (i..e. for eltypes not otherwise handled) to either throw an informative error or provide a time-domain implementation that only requires `+` and `*` be defined for the elements. Maybe throw an error for now and leave time-domain convolution for a future improvement.

Reply:
@martinholters I think it's better to provide time-domain convolution as a seperate function from `conv`, since the user may want to be able to choose between time-domain and FFT even if their arrays are numeric.  If we're to provide two implementations both under `conv` I think we should provide an explicit flag parameter for specifying the algorithm.  It seems strange to me that nominally different eltypes like `Float32` and `Gray{Float32}` can invoke completely different algorithms.

A better solution to the `Gray` problem in my opinion is something like a function `LinearAlgebra.blasfloat()` that is guaranteed to return a naked `BlasFloat`, unlike `float()` which may return a still wrapped type.

Also there is a package that does time-domain convolution: [aamini/FastConv.jl](aamini/FastConv.jl)

Reply:
@fghzxm allowing the user to specify the algorithm seems like an excellent idea. I personally like what ImageFiltering did for this, make a flag that is a type to allow dispatching to eliminate branches in some conditions (e.g. if the element type is not compatible with fourier transforms). Thanks for the suggestion with `blasfloat`, seems like a good solution. In terms of a time-domain convolution, I'll check out FastConv.jl. My impression last time I looked was that it was not actively maintained, but it seems like there's been some recent activity. Despite this, there are still some signs that it's not that up to date, e.g. it uses `conv2`, which was formerly in Julia base, then migrated to DSP.jl, and then deprecated. However, FastConv does not use DSP.jl. I'm not sure it would work atm. Also, looking at the code in FastConv, it looks like N dimensional case it's using `@nloops`, which has been soft-replaced by `CartesianIndices` etc.

DSP.jl already has fast time domain convolution for 1D inputs as part of the `filt` function, which could be easily adapted for general convolution. I'll try to add a time-domain branch of `conv` sometime soon for the 1D case, and think about how to do the ND case, and ideally also support custom array indexing. As was discussed in great detail in #355, I think the performance savings might be minimal, but perhaps a better reason to add time-domain convolution is to support convolution of arrays with any element type that has `+` and `*` defined, as @martinholters suggested. It could then act as a fall-back for inputs that are incompatible with fourier methods.